lib/http_aws_sigv4.c

```diff
@@ -408,6 +411,54 @@ CURLcode Curl_output_aws_sigv4(struct Curl_easy *data, bool proxy)
     }
   }
 
+  /* AWS S3 requires a x-amz-content-sha256 header, and supports special
+   * values like UNSIGNED-PAYLOAD */
+  sign_as_s3 = (strcasecompare(provider0, "aws") &&
+                strcasecompare(service, "s3"));
+
+  payload_hash = parse_content_sha_hdr(data, provider1, &payload_hash_len);
+
+  if(payload_hash) {
+    /* Use payload hash from existing content-sha256 header */
+  }
+  else if(sign_as_s3) {
+    /* Use UNSIGNED-PAYLOAD for the payload hash and content-sha256 header */
+    char *header = NULL;
+    struct curl_slist *head = NULL;
+
+    payload_hash_len = sizeof(S3_UNSIGNED_PAYLOAD) - 1;
```

Comment:
Can we add a DEBUGASSERT() here that verifies that `hash_len` is shorter than `sizeof(sha_hex)` ?